File: rag/local/retrieval.py
from __future__ import annotations

import hashlib
import json
import os
import sqlite3
import unicodedata
from dataclasses import dataclass
from pathlib import Path
from threading import Lock
from typing import Any

import faiss
import numpy as np
import torch
from FlagEmbedding import BGEM3FlagModel
from transformers import AutoModelForSequenceClassification, AutoTokenizer


PROJECT_ROOT = Path(__file__).resolve().parents[2]
RUNTIME_DIR = Path(os.environ.get("RAG_RUNTIME_DIR", PROJECT_ROOT / "rag" / "local" / "runtime"))
EMBEDDING_MODEL = "BAAI/bge-m3"
RERANKER_MODEL = "BAAI/bge-reranker-base"
INDEX_SCHEMA_VERSION = "local-rag-v1"
RRF_K = 60


@dataclass(frozen=True)
class Evidence:
    id: str
    locale: str
    source_title: str
    title_path: str
    content: str
    url: str
    evidence_level: str
    index_version: str
    hash: str
    score: float

    def public_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "locale": self.locale,
            "sourceTitle": self.source_title,
            "titlePath": self.title_path,
            "content": self.content,
            "url": self.url,
            "evidenceLevel": self.evidence_level,
            "indexVersion": self.index_version,
            "hash": self.hash,
            "score": self.score,
        }


def tokenize(value: str) -> list[str]:
    normalized = " ".join(unicodedata.normalize("NFKC", value).lower().split())
    latin: list[str] = []
    current = ""
    for character in normalized:
        if character.isascii() and (character.isalnum() or character in ".+*-"):
            current += character
        else:
            if len(current) >= 2:
                latin.append(current)
            current = ""
    if len(current) >= 2:
        latin.append(current)

    han_runs: list[str] = []
    run = ""
    for character in normalized:
        if "\u3400" <= character <= "\u9fff":
            run += character
        elif run:
            han_runs.append(run)
            run = ""
    if run:
        han_runs.append(run)

    stop = {"的", "了", "和", "与", "是", "在", "为", "及", "中"}
    han = [character for text in han_runs for character in text if character not in stop]
    han.extend(text[index : index + 2] for text in han_runs for index in range(max(0, len(text) - 1)))
    return list(dict.fromkeys([*latin, *han]))


def cache_key(chunk: dict[str, Any]) -> str:
    source = "\n".join(
        [INDEX_SCHEMA_VERSION, EMBEDDING_MODEL, str(chunk["hash"]), str(chunk["content"])]
    )
    return hashlib.sha256(source.encode("utf-8")).hexdigest()


class LocalRetriever:
    def __init__(self, runtime_dir: Path = RUNTIME_DIR) -> None:
        self.runtime_dir = runtime_dir
        self.index_path = runtime_dir / "faiss.index"
        self.ids_path = runtime_dir / "faiss-ids.json"
        self.database_path = runtime_dir / "knowledge.sqlite"
        self._embedding_model: BGEM3FlagModel | None = None
        self._reranker: AutoModelForSequenceClassification | None = None
        self._reranker_tokenizer: Any | None = None
        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self._index: faiss.Index | None = None
        self._ids: list[str] = []
        self._records: dict[str, dict[str, Any]] = {}
        self._lock = Lock()

    @property
    def ready(self) -> bool:
        return self.index_path.exists() and self.ids_path.exists() and self.database_path.exists()

    def warm(self) -> None:
        if not self.ready:
            raise RuntimeError("local_index_not_built")
        with self._lock:
            if self._index is None:
                self._index = faiss.read_index(str(self.index_path))
                self._ids = json.loads(self.ids_path.read_text(encoding="utf-8"))
                self._records = self._load_records()
            if self._embedding_model is None:
                self._embedding_model = BGEM3FlagModel(
                    EMBEDDING_MODEL,
                    use_fp16=True,
                    local_files_only=True,
                )
            if self._reranker is None:
                self._reranker_tokenizer = AutoTokenizer.from_pretrained(
                    RERANKER_MODEL,
                    local_files_only=True,
                )
                self._reranker = AutoModelForSequenceClassification.from_pretrained(
                    RERANKER_MODEL,
                    local_files_only=True,
                ).to(self._device).eval()
                if self._device.type == "cuda":
                    self._reranker.half()

    def health(self) -> dict[str, Any]:
        return {
            "ready": self.ready and self._index is not None and self._embedding_model is not None and self._reranker is not None,
            "indexLoaded": self._index is not None,
            "embeddingLoaded": self._embedding_model is not None,
            "rerankerLoaded": self._reranker is not None,
            "chunkCount": len(self._ids),
            "indexSchemaVersion": INDEX_SCHEMA_VERSION,
            "embeddingModel": EMBEDDING_MODEL,
            "rerankerModel": RERANKER_MODEL,
        }

    def retrieve(self, question: str, locale: str) -> list[Evidence]:
        self.warm()
        with self._lock:
            dense = self._rank_dense(question)
            keyword = self._rank_fts(question)
            fused = reciprocal_rank_fusion([dense, keyword])
            candidates = [
                self._to_evidence(chunk_id, score)
                for chunk_id, score in fused[:12]
                if chunk_id in self._records
            ]
            if not candidates:
                return []
            return self._rerank(question, candidates)[:6]

    def _load_records(self) -> dict[str, dict[str, Any]]:
        connection = sqlite3.connect(self.database_path)
        connection.row_factory = sqlite3.Row
        try:
            rows = connection.execute(
                "SELECT id, locale, source_title, title_path, content, url, evidence_level, index_version, hash FROM chunks"
            ).fetchall()
            return {str(row["id"]): dict(row) for row in rows}
        finally:
            connection.close()

    def _rank_dense(self, question: str) -> list[str]:
        assert self._embedding_model is not None
        assert self._index is not None
        vector = self._embedding_model.encode([question], batch_size=1, max_length=1024)["dense_vecs"]
        query = np.asarray(vector, dtype=np.float32)
        faiss.normalize_L2(query)
        _, indices = self._index.search(query, min(12, len(self._ids)))
        return [self._ids[index] for index in indices[0] if 0 <= index < len(self._ids)]

    def _rank_fts(self, question: str) -> list[str]:
        terms = tokenize(question)[:16]
        if not terms:
            return []
        match = " OR ".join(f'"{term.replace(chr(34), chr(34) * 2)}"' for term in terms)
        connection = sqlite3.connect(self.database_path)
        try:
            rows = connection.execute(
                "SELECT id FROM chunks_fts WHERE chunks_fts MATCH ? ORDER BY bm25(chunks_fts) LIMIT 12",
                (match,),
            ).fetchall()
            return [str(row[0]) for row in rows]
        except sqlite3.OperationalError:
            return []
        finally:
            connection.close()

    def _to_evidence(self, chunk_id: str, score: float) -> Evidence:
        record = self._records[chunk_id]
        return Evidence(
            id=chunk_id,
            locale=str(record["locale"]),
            source_title=str(record["source_title"]),
            title_path=str(record["title_path"]),
            content=str(record["content"]),
            url=str(record["url"]),
            evidence_level=str(record["evidence_level"]),
            index_version=str(record["index_version"]),
            hash=str(record["hash"]),
            score=score,
        )

    def _rerank(self, question: str, candidates: list[Evidence]) -> list[Evidence]:
        assert self._reranker is not None
        assert self._reranker_tokenizer is not None
        passages = [f"{item.title_path}\n{item.content}" for item in candidates]
        encoded = self._reranker_tokenizer(
            [question] * len(passages),
            passages,
            padding=True,
            truncation=True,
            max_length=1024,
            return_tensors="pt",
        )
        encoded = {key: value.to(self._device) for key, value in encoded.items()}
        with torch.inference_mode():
            scores = self._reranker(**encoded).logits.reshape(-1).float().cpu().tolist()
        return sorted(
            [Evidence(**{**item.__dict__, "score": float(score)}) for item, score in zip(candidates, scores)],
            key=lambda item: item.score,
            reverse=True,
        )


def reciprocal_rank_fusion(rankings: list[list[str]]) -> list[tuple[str, float]]:
    scores: dict[str, float] = {}
    for ranking in rankings:
        for index, chunk_id in enumerate(ranking):
            scores[chunk_id] = scores.get(chunk_id, 0.0) + 1.0 / (RRF_K + index + 1)
    return sorted(scores.items(), key=lambda item: item[1], reverse=True)

File: worker/rag/retrieval.ts
import corpusJson from "../../rag/corpus/public-knowledge.json";
import type { Evidence, KnowledgeChunk, Locale, RagEnv } from "./types";

const EMBEDDING_MODEL = "@cf/baai/bge-m3";
const RERANKER_MODEL = "@cf/baai/bge-reranker-base";
const RRF_K = 60;
const corpus = corpusJson as KnowledgeChunk[];
const byId = new Map(corpus.map((chunk) => [chunk.id, chunk]));
const searchable = corpus.map((chunk) => ({
  chunk,
  tokens: new Set(tokenize(`${chunk.titlePath} ${chunk.sourceTitle} ${chunk.content}`)),
  titleTokens: new Set(tokenize(`${chunk.titlePath} ${chunk.sourceTitle}`)),
}));

export async function retrieveEvidence(question: string, locale: Locale, env: RagEnv): Promise<Evidence[]> {
  const rankings = await Promise.all([
    Promise.resolve(rankStatic(question, locale)),
    rankD1(question, env),
    rankDense(question, env),
  ]);

  const fused = reciprocalRankFusion(rankings.filter((ranking) => ranking.length > 0));
  const candidates = fused
    .map(({ id, score }) => {
      const chunk = byId.get(id);
      return chunk ? { ...chunk, score } : null;
    })
    .filter((item): item is Evidence => item !== null)
    .slice(0, 12);

  return rerank(question, candidates, env);
}

function rankStatic(question: string, locale: Locale) {
  const queryTokens = [...new Set(tokenize(question))];
  if (!queryTokens.length) return [];

  return searchable
    .map(({ chunk, tokens, titleTokens }) => {
      let score = 0;
      for (const token of queryTokens) {
        if (tokens.has(token)) score += token.length > 2 ? 2 : 1;
        if (titleTokens.has(token)) score += 2.5;
      }
      if (chunk.locale === locale) score += 0.2;
      if (normalize(chunk.content).includes(normalize(question))) score += 8;
      return { id: chunk.id, score };
    })
    .filter((item) => item.score >= 1.2)
    .sort((a, b) => b.score - a.score)
    .slice(0, 12)
    .map((item) => item.id);
}

async function rankD1(question: string, env: RagEnv): Promise<string[]> {
  if (!env.DB) return [];
  const terms = tokenize(question).slice(0, 16);
  if (!terms.length) return [];

  const query = terms.map((term) => `"${term.replaceAll('"', '""')}"`).join(" OR ");
  try {
    const result = await env.DB.prepare(
      `SELECT knowledge_chunks.id AS id
       FROM knowledge_chunks_fts
       JOIN knowledge_chunks ON knowledge_chunks.rowid = knowledge_chunks_fts.rowid
       WHERE knowledge_chunks_fts MATCH ?1 AND knowledge_chunks.public = 1
       ORDER BY bm25(knowledge_chunks_fts)
       LIMIT 12`,
    )
      .bind(query)
      .all<{ id: string }>();
    return result.results.map((row) => row.id);
  } catch {
    return [];
  }
}

async function rankDense(question: string, env: RagEnv): Promise<string[]> {
  if (!env.AI || !env.VECTOR_INDEX) return [];

  try {
    const embedded = (await env.AI.run(EMBEDDING_MODEL, { text: [question] })) as {
      data?: number[][];
    };
    const vector = embedded.data?.[0];
    if (!vector) return [];
    const result = await env.VECTOR_INDEX.query(vector, {
      topK: 12,
      returnMetadata: true,
      filter: { public: true },
    });
    return result.matches.filter((match) => byId.has(match.id)).map((match) => match.id);
  } catch {
    return [];
  }
}

function reciprocalRankFusion(rankings: string[][]) {
  const scores = new Map<string, number>();
  for (const ranking of rankings) {
    ranking.forEach((id, index) => {
      scores.set(id, (scores.get(id) ?? 0) + 1 / (RRF_K + index + 1));
    });
  }
  return [...scores.entries()]
    .map(([id, score]) => ({ id, score }))
    .sort((a, b) => b.score - a.score);
}

async function rerank(question: string, candidates: Evidence[], env: RagEnv): Promise<Evidence[]> {
  if (!env.AI || candidates.length < 2) return candidates.slice(0, 6);

  try {
    const result = (await env.AI.run(RERANKER_MODEL, {
      query: question,
      contexts: candidates.map((item) => ({ text: `${item.titlePath}\n${item.content}` })),
    })) as { response?: Array<{ id?: number; score?: number }> };
    const scores = new Map(
      (result.response ?? []).map((item, index) => [item.id ?? index, item.score ?? 0]),
    );
    return candidates
      .map((item, index) => ({ ...item, score: scores.get(index) ?? item.score }))
      .sort((a, b) => b.score - a.score)
      .slice(0, 6);
  } catch {
    return candidates.slice(0, 6);
  }
}

function tokenize(value: string) {
  const normalized = normalize(value);
  const latin = normalized.match(/[a-z0-9][a-z0-9.+*-]{1,}/g) ?? [];
  const hanRuns = normalized.match(/[\p{Script=Han}]+/gu) ?? [];
  const han = hanRuns.flatMap((run) => {
    const characters = [...run];
    return [
      ...characters.filter((character) => !/[的了和与是在为及中]/.test(character)),
      ...characters.slice(0, -1).map((character, index) => character + characters[index + 1]),
    ];
  });
  return [...latin, ...han];
}

function normalize(value: string) {
  return value.normalize("NFKC").toLocaleLowerCase().replace(/\s+/g, " ").trim();
}
